docs/reference/esql/functions/types/bucket.asciidoc

```diff
@@ -6,14 +6,6 @@
 |===
 field | buckets | from | to | result
 datetime | integer | datetime | datetime | datetime
-datetime | integer | datetime | keyword | datetime
-datetime | integer | datetime | text | datetime
-datetime | integer | keyword | datetime | datetime
-datetime | integer | keyword | keyword | datetime
-datetime | integer | keyword | text | datetime
-datetime | integer | text | datetime | datetime
-datetime | integer | text | keyword | datetime
-datetime | integer | text | text | datetime
```

Comment:
This is interesting. We no longer tell folks they can use strings here, only dates. But maybe a human is ok because the examples will contain string literals.

Reply:
Is this correct, though? The function does accept the string types (the `foldToLong()` logic remained), but generally, wouldn't we have to rather extend the function infos of scalar/arithmetic functions to include these types, now that we autocast?

Reply:
Currently the implicit casting is done at resolution phase. We can do auto casting if the string literals are provided as input to the bucket function directly, e.g.
BUCKET(date, 20, "1985-01-01T00:00:00Z", "1986-01-01T00:00:00Z"). --> auto casting convert from and to to datetime

However, if the string literals are provided as input to the bucket function through a variable, auto casting does not apply, e.g.
| EVAL bucket_start = "1999-01-01T00:00:00Z". --> auto casting does not apply
| EVAL bucket_end = NOW()
| EVAL bucket = AUTO_BUCKET(hire_date, 5, bucket_start, bucket_end)

This is the main reason that there are still code inside to function to do the conversion, as the foldable eval propagation happens at logical planner. We have some options to make auto casting works better if needed, like:
- apply the implicit casting rule in logical planner after foldable eval propagation
- apply foldable eval propagation in resolution phase (I brought up this idea before, probably not a good one)

There are some follow up work related to the auto casting. A lot(perhaps most) of the rules in analyzer and logical planner are related to query transformation, there are chances that we can rearrange the rules to make resolution and query transformation work more effectively.

Reply:
As to whether this is right or not, I dunno. You certainly can only put date-like-strings in it. And now that we support implicit conversion for these literals everywhere where this list says "date" it means "date or date-like-string". I dunno.

Reply:
> This is the main reason that there are still code inside to function to do the conversion

Sure, that's fine. I was mostly commenting on adjusting the function info to remove the `keyword` and `text` types as accepted by the function (and the docs changes that ensued).